### DISCUSSION TITLE
[Question]: How to know strategy to solve given formula?

### DISCUSSION THREAD

**Discussion by hagozaebii** [cvc5/cvc5#10205] [General]
hello

As far as I know, there are various heuristics for solving formulas in cvc5(and other solvers), and I know that these are applied appropriately to the formula. These are expressed through a language called tactic, and the final sequence of heuristics used is known to be called strategy. I'm curious about the following:

1. Is there a way to see the strategy used when solving the formula in cvc5?
2. Is there a way to identify the heuristic being used while solving the formula in cvc5 in runtime?
3. Also, is there any method to know them using cvc5 base in python?

**Comment by martin-cs** [comment]
The terminology you use (tactics and strategy) are more commonly used when talking about interactive theorem provers (ITPs) or proof assistants.  There tactics and groups of tactics are very much part of the user-interface to the tool.

cvc5 does make uses of many heuristics, some are design level (assumptions about the relative frequency of operators, combinations of theories, etc.), some are algorithmic level (eager vs. lazy handling of different parts of the process, preprocessing and rewriting passes, approximations, VSIDS, etc.), some are configurable from the user level (the large number of options and switches).  These don't really map on to the same idea of tactics and strategy that ITP tools have.

As such, your questions are a little hard to answer.
1. The tracing infrastructure gives some insight into what is happening inside the solver.  It is hard to say which parts of this is "strategy".
2. Probably not in the way you would like.  Often there will be many of these running at the same time, layered and interacting.  You can look at the command-line flags used but I suspect this is not really what you want.
3. Not as far as I know.  The point of automatic theorem provers like cvc5 is that the majority of users *don't* have to know or care about what heuristics are being used.

**Comment by hagozaebii** [comment]
@martin-cs Thanks for your kind answer! If you are okay, I wanna ask you something more.

First of all, I want to clarify the meaning of `strategy` that I used. The meaning of `strategy` that I used is sequence of `heuristic` that cvc5 used to solve formulas. I used the terminology with the same meaning in the paper(https://files.sri.inf.ethz.ch/website/papers/fastsmt-2018.pdf).
And, by this question, I wanted to know `how to get sequence of heuristics` that used while solving. For example, z3 can get it with `-v:10` option. The z3 result with this option is as follows:
```
(simplifier :num-exprs 83 :num-asts 416 :time 0.00 :before-memory 18.07 :after-memory 18.07)
(propagate-values :num-exprs 67 :num-asts 427 :time 0.00 :before-memory 18.07 :after-memory 18.07)
(solve-eqs :num-exprs 20 :num-asts 428 :time 0.00 :before-memory 18.07 :after-memory 18.07)
(:solve-eqs-elim-vars 17
 :solve-eqs-steps     52)
(elim-uncnstr :num-exprs 22 :num-asts 441 :time 0.00 :before-memory 18.07 :after-memory 18.07)
(:elim-unconstrained 1)
(:reduced-funcs 1)
(reduce-args :num-exprs 21 :num-asts 444 :time 0.00 :before-memory 18.07 :after-memory 18.07)
(reduce-bv-size :num-exprs 21 :num-asts 444 :time 0.00 :before-memory 18.07 :after-memory 18.07)
(max-bv-sharing :num-exprs 21 :num-asts 444 :time 0.00 :before-memory 18.07 :after-memory 18.07)
(ackermannize_bv :num-exprs 21 :num-asts 444 :time 0.00 :before-memory 18.07 :after-memory 18.07)
(simplifier :num-exprs 19 :num-asts 451 :time 0.00 :before-memory 18.07 :after-memory 18.07)
(propagate-values :num-exprs 19 :num-asts 451 :time 0.00 :before-memory 18.07 :after-memory 18.07)
(solve-eqs :num-exprs 18 :num-asts 451 :time 0.00 :before-memory 18.07 :after-memory 18.07)
(:solve-eqs-elim-vars 1
 :solve-eqs-steps     1)
(elim-uncnstr :num-exprs 14 :num-asts 469 :time 0.00 :before-memory 18.07 :after-memory 18.07)
(:elim-unconstrained 3)
(reduce-bv-size :num-exprs 14 :num-asts 469 :time 0.00 :before-memory 18.07 :after-memory 18.07)
(simplifier :num-exprs 8 :num-asts 464 :time 0.00 :before-memory 18.07 :after-memory 18.07)
(simplifier :num-exprs 8 :num-asts 462 :time 0.00 :before-memory 18.07 :after-memory 18.07)
(max-bv-sharing :num-exprs 8 :num-asts 462 :time 0.00 :before-memory 18.07 :after-memory 18.07)
(ackermannize_bv :num-exprs 8 :num-asts 462 :time 0.00 :before-memory 18.07 :after-memory 18.07)
(bit-blast :num-exprs 4 :num-asts 464 :time 0.00 :before-memory 18.07 :after-memory 18.07)
(simplifier :num-exprs 4 :num-asts 464 :time 0.00 :before-memory 18.07 :after-memory 18.07)
(solve-eqs :num-exprs 0 :num-asts 463 :time 0.00 :before-memory 18.07 :after-memory 18.07)
(:solve-eqs-elim-vars 2
 :solve-eqs-steps     2)
sat
```
I want a similar result in cvc5 as well. May I know how to get a similar result in cvc5?

Thanks for your kindness always!

**Reply by martin-cs** [reply]
`cvc5` doesn't really work like this.  This is an ITP-style feature that was added to z3, I guess while Leonardo's thinking was heading towards the design of Lean.  We don't have a way of "applying a given tactic"; that's how ITPs work not ATPs.

You could use the proof, which, literally tells you what steps were needed.  If you produce this as a series of assertions then I guess you could replay it in some kind of way?

There is some ongoing work in the community in producing modular solvers which might do what you want but it requires some deeper work within the architecture of a solver.

I'm happy to discuss this more but it might be easier via Teams or something interactive.

**Comment by hagozaebii** [comment]
@martin-cs Thanks for your kind answer always :)
